eclipse-theia/theia
# Improve Plugin Preference Access

#### What it does


Fixes #10004
Depends on #11331
This PR updates various aspects of preference access and setting from plugins.
 - Adjusts PreferenceRegistry to use Monaco code for `Configuration` and `ConfiguratioModel`.
 - Enables access and setting of language-overridden preferences from plugins according to VSCode API (#10004)
 - Updates `affectsConfiguration` check to respect override options (language and URI)


#### How to test


0. Assert that the tests pass - most of the functionality related to preference access and `affectsConfiguration` has new tests.
0. Install [this plugin](https://github.com/eclipse-theia/theia/files/9067216/extension-preference-access-0.0.1.zip) ([source](https://github.com/colin-grant-work/demo-extensions/tree/main/extension-preference-access))
1. Open a single-root workspace.
2. Run the command `Listen for config changes`
3. Add and remove tasks, observe that a toast informs you that you have done so. 
4. Delete your `tasks.json` file.
5. Observe that a toast registers for changes to `tasks` and `tasks.tasks`
> On `master`, the deletion of the file would not be registered by plugins as a change in the section.
6. Modify the value of `editor.fontSize` in your workspace; modify it for language overrides, including `[rust]`
7. Observe that your changes are accurately reported.
8. Clear the values of `editor.fontSize` in your workspace.
> Run the command 'Stop listening for config changes' now if you want to see fewer toasts in the next step.
9. Run the command 'Run config test (no language override)'
10. It should modify the value for `editor.fontSize` in your workspace and report that the effective value has changed for non-overridden and for `rust`.
11. Run the command 'Run config test (with language override)'
12. Same as 10., but with different numerical values.
13. Run the command 'Run config test (with forced language override)'
14. It should create an overridden value for `editor.fontSize` for the `rust` language, and the report should show that only `rust` and not un-overridden `editor.fontSize` was affected
15. Run the command 'Run config test (with language override)' again.
16. This time, because an override already exists for `rust` after (13), you should see a toast similar to 14 - only the value for `rust` should be modified.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
